src/utils/series.ts

```diff
@@ -1,85 +1,103 @@
-  }
-
-  return series.episodes;
+/**
+ * Get a playlist with episodes based on the selected filter
+ */
+export const filterSeries = (playlist: Playlist, episodes: EpisodesWithPagination[] | undefined): Playlist => {
+  // Get a flattened list of episodes and return it as part of the playlist
+  return { ...playlist, playlist: episodes?.flatMap((e) => e.episodes) || [] };
 };
 
-/** Next episode to show after video is complete */
-export const getNextEpisode = (series: Series, media: PlaylistItem): string => {
-  // To handle array elements
-  const episodes = getSeriesEpisodes(series);
+/**
+ * Retrieve a first episode of the next season
+ */
+const getNextSeasonEpisode = async (seasonNumber: number, series: Series | undefined) => {
+  const seasons = series?.seasons;
+
+  // The seasons is the last one in case there is only one season available or it is the same as the last season in the seasons array
+  const isLastSeason = seasons?.length === 1 || seasons?.[seasons?.length - 1]?.season_number === seasonNumber;
 
-  const episodeIndex = episodes?.findIndex((el) => el.media_id === media.mediaid);
+  if (isLastSeason) {
+    return;
+  } else {
+    // Need to know the order of seasons to choose the next season number
+    const hasAscendingOrder = (Number(seasons?.[1]?.season_number) || 0) > (Number(seasons?.[0]?.season_number) || 0);
+    const nextSeasonNumber = hasAscendingOrder ? seasonNumber + 1 : seasonNumber - 1;
 
-  // If we have the last episode => we leave the same video
-  if (episodeIndex === episodes.length - 1) {
-    return media.mediaid;
+    return (await getSeasonWithEpisodes(series?.series_id, nextSeasonNumber, 0, 1))?.episodes?.[0];
   }
+};
 
-  return episodes[episodeIndex + 1]?.media_id;
+/**
+ * Get a next episode of the selected season / episodes list
+ */
+const getNextEpisode = async (seasonNumber: number, seriesId: string, pageWithEpisode: number) => {
+  if (seasonNumber) {
+    return (await getSeasonWithEpisodes(seriesId, seasonNumber, pageWithEpisode, 1))?.episodes?.[0];
+  } else {
+    return (await getEpisodes(seriesId, pageWithEpisode, 1))?.episodes?.[0];
+  }
 };
 
-export const getNextItem = (item: PlaylistItem | undefined, series: Series | undefined, seriesPlaylist: Playlist | undefined): PlaylistItem | undefined => {
-  if (!item || !seriesPlaylist) return;
+const getNewFlowNextEpisode = async (series: Series | undefined, episodeMetadata: EpisodeMetadata | undefined) => {
+  if (!episodeMetadata || !series) {
+    return;
+  }
+
+  const seasonNumber = Number(episodeMetadata.seasonNumber);
+  const episodeNumber = Number(episodeMetadata.episodeNumber);
 
-  if (series) {
-    const nextEpisodeId = getNextEpisode(series, item);
+  // Get initial data to collect information about total number of elements
+  const { episodes, pagination } =
+    seasonNumber !== 0 ? await getSeasonWithEpisodes(series?.series_id, seasonNumber, 0) : await getEpisodes(series?.series_id, 0);
```

Comment:
Is there a way to rewrite this so that we make better use of caching here? If the seasons or episodes are already loaded, we don't want to make another BE call and vice versa.

Reply:
The problem in this case will be refetching of 'next episode' when switching the season selected. In case we pass current episodes as a dependency, then we will refetch the data every time the array changes.